### ISSUE TITLE
Construction of DisjointClasses Axioms with exactly one distinct member. 

### ISSUE THREAD

**Issue by sesuncedu** [owlcs/owlapi#306]
This is the basic  issue in #151 (tl;dr - arguments to disjointness axioms should not have duplicates removed). 

For version 5, this should be reflected in the API; possibly by lowering the argument types of the factory methods  to Collection<>. 

For versions 3 and 4 it is tricky to solve the full problem, but there are some hacks that are plausible. 
Assuming that the published API cannot change, the Disjointness axioms builders must take arguments of type Set<>, return objects of type DisjointBlahAxiom, with accessors that must return values of type Set<>. 

 For the simplest  case like DisjointClasses(:x :x),  with all arguments identical, data factory methods like getOWLDisjointClassesAxiom would be called with a set containing just a single member.

This can be treated as being equivalent to DisjointClasses(:x owl:Thing) (i.e. :x === owl:Nothing). This is hacky, but it keeps the API the same.  

A DisjointClasses with multiple arguments, some of which are duplicated, would have to be deliberately constructed, and would be hard to detect. The situation can't occur when parsing RDF formats, since everything is pair-shaped.  

DisjointProperties might raise problems with profiles, but DisjointClasses ought to be safe.

**Comment by ykazakov** [comment]
Sorry fo commenting to an old ticket (which is a follow up of #151), but it is still not clear to me whether the problems with duplicate classes in disjoint axioms suppose to be resolved or it will not be fixed?

I did a quick test with OWL API 5.1.4:

```
OWLOntologyManager man = OWLManager.createOWLOntologyManager();
		OWLDataFactory f = man.getOWLDataFactory();
		OWLClass A = f.getOWLClass(IRI.create("A"));
		OWLAxiom ax = f.getOWLDisjointClassesAxiom(A, A);
		System.err.println(ax);
		OWLClass B = f.getOWLClass(IRI.create("B"));
		ax = f.getOWLDisjointClassesAxiom(A, B, A);
		System.err.println(ax);
```

The first disjoint axiom is printed as:
```
DisjointClasses(Annotation(Annotation(rdfs:comment "Generated by the OWL API (version 5.1.4.2018-01-04T07:01:37Z) https://github.com/owlcs/owlapi/ on 3/24/18 10:24 PM"^^xsd:string) rdfs:comment "DisjointClasses(<A>) replaced by DisjointClasses(<A> owl:Thing)"^^xsd:string) <A> owl:Thing)
```
which is equivalent to `DisjointClasses(<A> <A>)`, although with some noise.
However the second disjoint axiom is not equivalent to the original one:
```
DisjointClasses(<A> <B>)
```
Why duplicate arguments still get swallowed in OWL API 5 where the members of axioms are no longer sets?

**Comment by ignazio1977** [comment]
The input is a Collection but the axiom treats these as sets and remove duplication, as this is the intent of the specs - as recognized by all who commented on the original thread.

Treating `Disjoint(A, A)` in a special way (with added intentional noise) was a hack to satisfy the need for some systems that shall remain unnamed to use this idiom (which is /not/ in agreement with the specs) for whatever purpose they see fit.

There's a separate argument about whether the specs meant the input to axioms to be lists, sets, bags or something else. The argument was raised on the public-owl mailing list; the agreement so far is that the specs, whatever the intent of the authors, describe a set-like behaviour. There seemed to be dissent on whether this was to be changed or not. I am not aware of the argument having progressed.

I'm not happy with the hack currently in place - I think the invalid axioms should just be dropped. However, I'm not eager to reopen that can of worms.

On a related issue (why are the arguments to OWLDataFactory not Sets?) it turned out that there are plenty of usages of OWLDataFactory where axioms are created for a variety of purposes (e.g., temporary axioms in applications such as clustering and reasoning) where forcing the interface to be Sets did cause a measurable performance hit without providing any extra features (the collection internal to the axiom implementation itself is always a List because it's the cheapest structure that guarantees order, given the immutability of the axiom instance - well, cheapest next to an array, but using arrays caused cumbersomeness on a different level, this time Java's fault, because arrays of generic types can't be created and worked on as conveniently as lists.)
Navigating lists with streams also appears to be slightly faster than navigating sets; however that was a smaller issue.

**Comment by ignazio1977** [comment]
Short version: duplicates are removed, except when duplicate removal leaves the set with only one element.

**Comment by ykazakov** [comment]
Sorry for opening the old wound :-/.
@ignazio1977, can you give a link to the discussion on the owl mailing list you mentioned? This would be useful to understand the background.

To me the situation with the disjoint classes is still not clear. Is it a syntactic problem (a problem with serialisation, e.g., like #468), or is it a semantic problem?
Why duplicate single occurrence in DisjointClasses cannot be treated in the same way as, e.g., for ObjectIntersectionOf, which also requires two arguments? I just tested: 
```
System.out.println(f.getOWLObjectIntersectionOf(A, A));
```
It prints:  `ObjectIntersectionOf(<A>)`, not `ObjectIntersectionOf(<A> owl:Thing)`, which would be in accordance to this "hack".

Specifically, according to the OWL specs, should `DisjointClasses(<A> <A>)` imply that A is inconsistent or not? For what reason `DisjointClasses(<A> <A>)` is printed in OWL API as `DisjointClasses(<A> owl:Thing)` and not, say, `DisjointClasses(<A> owl:Nothing)`? The letter was suggested by @ignazio1977 before the ticket was closed. Is it a typo? To me, `DisjointClasses(<A> <A> <B>)` should be stronger than `DisjointClasses(<A> <A>)`, and it is counter-intuitive why A is inconsistent according to the second axiom but not according to the first one.

**Comment by ykazakov** [comment]
Thanks, @ignazio1977 
I think, these messages can be found in the archive, and also a comment by Michael Wessel:
https://lists.w3.org/Archives/Public/public-owl-comments/2014Apr/0000.html
Interesting reading, indeed.

To some up, does it simply mean that `DisjointClasses(<A> <A>)`, `DisjointClasses(<A> <A> <B>)`, `ObjectIntersectionOf(<A> <A>)`, etc., are all not valid OWL 2 expressions, and for that reason, their semantics is not defined, and for that reason, tools like OWL API can deal with these expressions as they see it fits best? In this case, treating arguments as lists should be consistent with OWL specification as the observed behaviour will be the same on *valid* OWL ontologies.

Further, I think there should be a (possibly switchable) warning that the expressions with singleton arguments are not valid. And also ontologies with duplicate occurrences in set positions are not valid. Otherwise these issues remain unnoticed and get accumulated. I recall, there are even ontologies which have something like `ObjectIntersectionOf()`.

Regarding your explanation with OpenCyc, it sounds like a problem with a specific ontology. Should this "hack" be really a part of the OWL API? Why not to simply fix the ontology or deal with this problem programmatically?

**Comment by ykazakov** [comment]
Also, I think treatment of incorrect OWL expressions should be consistent. It is not good to (silently) ignore expressions like `DisjointUnion(<A> <B>)` in the output of FSS, but print (invalid for the same exactly reason) expressions like `EquivalentClasses(<A> ObjectUnionOf(<B>))` (according to #468).

**Comment by ignazio1977** [comment]
> `DisjointClasses(<A> <A> <B>)`

When the input was a set, this was impossible - it would be reduced to `(<A>, <B>)`. This still happens, and I believe it has not caused unexpected behaviour (i.e., no one has attempted to use it to signify something like B is disjoint with owl:Nothing and A is equivalent to owl:Nothing, which is the only way to cast the intended RDF meaning in OWL axioms - far as I could tell).

On the other hand, the errata link contains a fix that escaped my earlier read - i.e., that `DisjointClasses(A, A)` actually be rendered as `SubClassOf(A, owl:Nothing)`. I assume this would entail that `DisjointClasses(A, A, B, C)` becomes `SubClassOf(A, owl:Nothing)` + `DisjointClasses(A, B, C)`. 

This is still bad in the OWLDataFactory interface, as each method specifies the returned axiom type, and therefore the axiom cannot be created as requested, making it programmatically impossible to solve the issue. At parsing and rendering, the situation is slightly better, as we could check beforehand and `Do the right thing`(tm), however it requires code duplication across parsers and renderers, so still a major headache.

`DisjointUnion` is also mentioned, which in retrospect is obvious but it escaped my notice in the last four years. I'm not sure what the disjoint union of one class would mean.

**Comment by ignazio1977** [comment]
> Regarding your explanation with OpenCyc, it sounds like a problem with a specific ontology. Should this "hack" be really a part of the OWL API? Why not to simply fix the ontology or deal with this problem programmatically?

:-) that was precisely my position when the issue was brought up. I got worn down.

**Comment by ykazakov** [comment]
Well, according to the linked discussion on OWL mailing list, `DisjointClasses(A, A)` should not be a valid axiom. @ignazio1977, where did you find that it should be equivalent to `SubClassOf(A, owl:Nothing)`? Disjoint union with one (or repeated) class expressions is similarly invalid. So what we are discussing, is how to best handle such invalid expressions in some consistent and predictable way. 

On the one hand, there should be a way in OWL API to check and/or warn about such invalid expressions. On the other hand, in most cases there is not much harm in having duplicate / singleton (maybe even empty) expressions, since the semantics will be the same as if duplicates removed. Except for disjointness axioms, which is the topic of the discussion here.

My suggestion is, that OWL API should preserve duplicates and the order of arguments. In particular, when loading `DisjointClasses(A, A)` (e.g., from an ontology in FSS), just load it as a DisjointClass axiom which outputs A two times when `classExpressions()` is called. So when printed (through FSS renderer) the output will be the same expression. This should now be possible since arguments are switched from sets to collections (and internally backed by lists). For valid ontologies (i.e., which do not have duplicates or singletons) there will be no observable difference. For invalid ontologies, e.g., those containing `DisjointClasses(A, A)`, we delicate it to the reasoner to handle this problem.

**Comment by ykazakov** [comment]
Some correction after further reading: it looks like `DisjointClasses(<A> <A> <B>)` is a correct expression in FSS syntax. According to the [OWL2 Structural Specification](https://www.w3.org/TR/owl2-syntax/#Structural_Specification): 

> Sets written in one of the exchange syntaxes (e.g., XML or RDF/XML) are not necessarily expected to be duplicate free. Duplicates should be eliminated when ontology documents written in such syntaxes are converted into instances of the UML classes of the structural specification.

So, `DisjointClasses(<A> <A> <B>)` should be converted to `DisjointClasses(<A> <B>)`, which does not imply inconsistency of A.
On the other hand, `DisjointClasses(<A> <A>)` is not a valid expression in FSS syntax since its conversion `DisjointClasses(<A>)` is not a valid OWL expression.

To sum up: 

1. Elimination of structurally equivalent expressions in set arguments seems to be necessary to comply with the structural specification. This is unfortunate since it requires more work and in almost all cases have no impact on reasoning results.
2. Handling of `DisjointClasses(<A> <A>)` as well as `DisjointClasses(<A>)` or `ObjectIntersectionOf(<A>)`, or `ObjectIntersectionOf()` is out of specification, and, I think, an appropriate warning should be issued in those cases.
3. Assuming that representing such out-of-spec cases is useful, I would prefer if `DisjointClasses(<A> <A>)` were loaded as `DisjointClasses(<A>)`, and the reasoner would be responsible to dealing with its meaning. The reasoner can decide, e.g., that in this case it should mean that A is unsatisfiable, or simply throw an exception. If necessary, one can add an annotation with the original axiom, which a reasoner may take into account.

**Comment by ignazio1977** [comment]
> where did you find that it should be equivalent to SubClassOf(A, owl:Nothing)

This was one of the suggested solutions in the errata, from Ian's last message in the email exchange above https://www.w3.org/2001/sw/wiki/OWL_Errata

**Comment by ignazio1977** [comment]
Trying to catch up here - might be worth starting one or more separate issue threads.

> DisjointClasses(<A> <A> <B>) should be converted to DisjointClasses(<A> <B>), which does not imply inconsistency of A

Yep, this is the current behaviour. Inconsistency of A, (if it were in the intention of the writer) is lost.

1. It is more work to remove duplication, but at least to the extent that OWLAPI is concerned the structural equivalence is reduced to the `OWLObject.equals()` implementation. This is already implemented and its runtime speed, although not above suspicion (plenty of issues opened and closed on that topic), is not a big deal. At reasoning level, detecting equivalent class expressions and figuring out they coappear in disjoint axioms is, if I recall correctly, something that needs doing anyway.

2. Agreed that it is out of specification. I believe that DisjointClasses(A,A) is the only case that does not result in an exception being thrown - this is the case where annotations on the output axioms are instead written. Happy to discuss what an `appropriate` warning should be - I've heard that particular turn of phrase before and you wouldn't believe how many different meanings of appropriate people consider appropriate for the same situation :-) currently parsing has a strict and loose mode (for RDF syntaxes at least), the level of appropriateness could be linked to the parsing mode.

3. I'm open to the possibility; not sure about the specifics.

**Comment by ignazio1977** [comment]
> OWL API should preserve duplicates and the order of arguments

Uhm, this is definitely something I'm uncomfortable with. Preserving duplication is a change of behaviour that will likely impact applications iterating over signatures; it also introduces interesting questions about axiom equality - duplicates in a signature do or do not cause otherwise identical axioms to be disequal?
Same applies to order; it used to be that order of collections in axioms and class expressions was not guaranteed, and that caused both confusion to users (normalizing input when the axiom is created is one thing, but the order of operands changing afterwards on a regular basis confuses users) and an impact on serialization - I can't find the exact log numbers but changing order causes minute, unnecessary changes in the text files produced by serialization, which cause often large patch sets for ontologies kept in versioning systems that do not know about RDF - e.g., Git.

So, I see more impact than benefit in this change, and it also counts as a large behavioural change. I don't think it would win many fans among the most assiduous API users.

**Comment by ykazakov** [comment]
> Yep, this is the current behaviour. Inconsistency of A, (if it were in the intention of the writer) is lost.

How come? This ticket is about the case of `DisjointClasses(<A>)` and currently is to converted to `DisjointClasses(<A> owl:Thing)`, which should imply inconsistency of A. B.t.w., an interesting case is what happens to `DisjointClasses(owl:Thing)`, which seems to be converted to just `DisjointClasses(owl:Thing)` with a long annotation, and does not force inconsistency of owl:Thing.

>  I believe that DisjointClasses(A,A) is the only case that does not result in an exception being thrown - this is the case where annotations on the output axioms are instead written.

You probably talk about exceptions thrown when *parsing* ontologies in serialisation formats? (B.t.w., illegal expressions like ObjectIntersectionOf(A A), ObjectUnionOf(A A), etc., should also parse without any notice). But I am talking about exceptions when trying to *create* illegal OWL objects by any means, e.g., using OWL factories. For these cases an [IllegalArgumentException](https://docs.oracle.com/javase/7/docs/api/java/lang/IllegalArgumentException.html) seems appropriate.

> Happy to discuss what an appropriate warning should be - I've heard that particular turn of phrase before and you wouldn't believe how many different meanings of appropriate people consider appropriate for the same situation :-)

What about producing a log message, e.g., "Warning: OWL expression: *** is illegal according to the OWL 2 specification!"  Does OWL API produce any log messages? It seems to depend on SLF4J.

> Uhm, this is definitely something I'm uncomfortable with. Preserving duplication is a change of behaviour that will likely impact applications iterating over signatures; it also introduces interesting questions about axiom equality - duplicates in a signature do or do not cause otherwise identical axioms to be disequal?

What the order of arguments has to do with iterating over signatures? You can still return the signature of an axiom as a set. You can still return the arguments of an axiom as a set (using the now deprecated OWL API 4 methods), but, in addition, you can return the arguments as a stream, in which the order and duplications could potentially be preserved. All applications that use OWL API 4 methods will behave exactly as before. OWL API 5 applications will have a choice which methods to use.

> Same applies to order; it used to be that order of collections in axioms and class expressions was not guaranteed, and that caused both confusion to users (normalizing input when the axiom is created is one thing, but the order of operands changing afterwards on a regular basis confuses users) and an impact on serialization - I can't find the exact log numbers but changing order causes minute, unnecessary changes in the text files produced by serialization, which cause often large patch sets for ontologies kept in versioning systems that do not know about RDF - e.g., Git.

I cannot follow this logic. Preserving duplicates and the order would somehow change the order in the serialisation? I think this will be exactly opposite: the expressions will be serialised in exactly the same way as they were parsed.

Regarding the equals() method of OWL API implemented as the "structural equivalence," I found it unfortunate that this choice has been decided for the developers. This, in particular, causes the above mentioned problems with the order of serialisation etc. For reasoners, the structural equivalence plays little importance. Yes, reasoners should be able to identify structurally equivalent objects, as they should be able, in general, to identify *any* equivalences, but they don't have to do that unless being explicitly asked for the same reason why they should not merge generally equivalent arguments.  E.g., from EquivalentClasses(A B C) a reasoner can conclude that A is subsumed by B without necessarily verifying structural equivalence of the arguments. ELK reasoner, for example, does not have a structural equivalence checker in its OWL model, and disjointness axioms are the only cases which force us to implement this.

**Comment by ignazio1977** [comment]
>> Yep, this is the current behaviour. Inconsistency of A, (if it were in the intention of the writer) is lost.

>How come? This ticket is about the case of DisjointClasses(<A>) and currently is to converted to DisjointClasses(<A> owl:Thing), which should imply inconsistency of A. B.t.w., an interesting case is what happens to DisjointClasses(owl:Thing), which seems to be converted to just DisjointClasses(owl:Thing) with a long annotation, and does not force inconsistency of owl:Thing.

Bad formatting of my previous message caused context to be lost.

Reprinting it:

>> `DisjointClasses(<A> <A> <B>)` should be converted to `DisjointClasses(<A> <B>)`, which does not imply inconsistency of A
>
> Yep, this is the current behaviour. Inconsistency of A, (if it were in the intention of the writer) is lost.

I was not referring to single input or single unique input.

**Comment by ignazio1977** [comment]
> I believe that DisjointClasses(A,A) is the only case that does not result in an exception being thrown - this is the case where annotations on the output axioms are instead written.

> You probably talk about exceptions thrown when parsing ontologies in serialisation formats? (B.t.w., illegal expressions like ObjectIntersectionOf(A A), ObjectUnionOf(A A), etc., should also parse without any notice). But I am talking about exceptions when trying to create illegal OWL objects by any means, e.g., using OWL factories. For these cases an IllegalArgumentException seems appropriate.

Parsers use `OWLDataFactory` to create axioms, so an exception thrown from `OWLDataFactory` would cause the parser to fail too. The parsing issue that originated this is not of this kind - the exception parsing DisjointClasses with only one argument comes from the Functional Syntax grammar, not OWLDataFactory.

I /think/ there's a check and exception being thrown in OWLDataFactory for singleton collections, but can't recall exactly. I agree there /should/ be an exception. Might need to be tied to something like he strict/loose parsing mode, if the current behaviour is not to throw exceptions.

**Comment by ignazio1977** [comment]
> What about producing a log message, e.g., "Warning: OWL expression: *** is illegal according to the OWL 2 specification!" Does OWL API produce any log messages? It seems to depend on SLF4J.

That's doable. Yes, OWLAPI produces plenty of log messages - which ones are actually visible depends on the calling code. OWLAPI does not impose a logging level or a logging appender - this approach for libraries seems to be the most common expectation. There's a massive discussion on Phil Lord's blog about this and what a sensible expectation is, if you fancy a digression.

**Comment by ignazio1977** [comment]
> What the order of arguments has to do with iterating over signatures? You can still return the signature of an axiom as a set.

Creating a set for each call is a massive performance hit. Workarounds of many sorts, such as a Set implementation backed by a List, have been in the OWL API since about 2010, in response to user complaints.

**Comment by ykazakov** [comment]
> Parsers use OWLDataFactory to create axioms, so an exception thrown from OWLDataFactory would cause the parser to fail too.
> I /think/ there's a check and exception being thrown in OWLDataFactory for singleton collections, but can't recall exactly. I agree there /should/ be an exception. Might need to be tied to something like he strict/loose parsing mode, if the current behaviour is not to throw exceptions.

The OWLDataFactory does not throw any exceptions for me. I can call, e.g., `getOWLObjectIntersectionOf()` with a singleton argument or even with the empty argument and nothing fails. So I think the parsing exceptions are issued by the parser code, not by OWLDataFactory. If the would be no way to create illegal OWL expressions then there would not be tickets like #46 or #468.

**Comment by ykazakov** [comment]
>  OWLAPI does not impose a logging level or a logging appender - this approach for libraries seems to be the most common expectation.

I do not understand. When issuing a log message (e.g., using SLF4J or LOG4J), one usually needs to use a logger object and chose an appropriate log level. E.g., logger.debug("foo") or logger.warn("bar"). Which of these messages are going to be displayed or saved in a file, of course, depends on the configuration of the client.

**Comment by ykazakov** [comment]
>> What the order of arguments has to do with iterating over signatures? You can still return the signature of an axiom as a set.

> Creating a set for each call is a massive performance hit. Workarounds of many sorts, such as a Set implementation backed by a List, have been in the OWL API since about 2010, in response to user complaints.

How can you avoid creating a set for the *signature* if the *arguments* of OWL objects are stored as set? Why would it be any different if arguments are stored, e.g., for lists? OWL expressions can be nested, so signature elements may appear in sub-expressions many times even if every argument of every operator is unique.

**Comment by ykazakov** [comment]
> I proposed preserving the parse order as a possible solution, but people had a bigger problem with axioms undergoing modifications - which means, an axiom replaced by a new axiom with new operands, or removed operands. Preserving the parsing order became much harder in such scenarios (it could have been preserved by passing the original axiom as part of the input to a factory, but this would have required changes in all the client code that modified axioms).

I the arguments of expressions are represented as lists, I do not see much difficulties in preserving the order. In fact, the users would be in full control of the order of arguments.

> Duplicate removal was never put in doubt - no one wanted the input to behave like a list or a multiset; they wanted it to behave like a sorted set.

Well, in 99.9% of the use cases arguments do not contain duplicates anyway. So there would be no difference between a list or a (manually) sorted set. So no one really cared. But occasionally, duplicates may appear and I am not sure it is sensible to always (silently) eliminate them. If the input contained duplicate arguments then maybe it was needed for something?

For example, we want to provide explanations of entailments in ELK as derivations (sequences of OWL axioms where every next axiom is derived from previous axioms by a simple rule). For this, we may need to use some tautological axioms, like SubClassOf(A A) or EquivalentClasses(ObjectIntersectionOf(A B) ObjectIntersectionOf(B A)) as initial axioms (if one wants to explain a derivation of a tautology, one always need to start with tautologies). The letter axiom would express that the conjunction is a commutative operation. However, OWL API would "swallow" the second argument of EquivalentClasses and our derivation would be broken. It is very difficult to predict when such things happen.
